src/price_data.py

```diff
@@ -326,3 +328,133 @@ def get_cost(self, tr: Union[Operation, SoldCoin], reference_coin: str = config.
         if isinstance(tr, SoldCoin):
             return price * tr.sold
         raise NotImplementedError
+    
+    def get_candles(self, start: int, stop: int, symbol: str) ->list:
+        if self.exchange.has['fetchOHLCV']:
+            sleep(self.exchange.rateLimit / 1000)  # time.sleep wants seconds
+            # get 2min before and after range
+            return self.exchange.fetch_ohlcv(symbol, '1m', start-1000*60*2, max(int((stop-start)/1000/60)+5, 1))
+        else:
+            logging.warning(
+                "fetchOHLCV not implemented on exchange, skipping priceloading using ohlcv")
+            raise None
+
+    def initialize_ccxt(self):
+        exchange_id = 'binance'
+        exchange_class = getattr(ccxt, exchange_id)
+        self.exchange = exchange_class()
+        self.markets = []
+        markets = self.exchange.fetch_markets()
+
+        for market in markets:
+            # may not apply for all exchanges, currently works for binance
+            # caches a list of all pairs on the exchange
+            self.markets.append(market["symbol"].split("/"))
+
+    def _get_binance_bulk_pair_list(self, reference_coin: str = config.FIAT, coin) -> list:
+
+        def get_pair(coin, reference_coin:str):
+
+            for market in self.markets:
+                if market[0] == coin and market[1] == reference_coin:
+                    return [coin, reference_coin, False]  # False=not inverted
+                elif market[1] == coin and market[0] == reference_coin:
+                    return [reference_coin, coin, True]  # True=inverted
+
+        pair = get_pair(coin, reference_coin)
+        if not pair:
+
+            for market in self.markets:
+                pair = get_pair(market[1], reference_coin)
+
+                if pair:
+                    if market[0] == coin:
+                        return [[market[0], market[1], False], pair]
+                    elif market[1] == coin:
+                        return [[market[1], market[0], True], pair]
+        else:
+            return [pair, ]
+
+    def _get_binance_bulk_pair_data(self, operations: list, symbol: str, invert: str=False) ->list:
+        timestamps = []
+        timestamppairs = []
+        data = []
+
+        for op in operations:
+            timestamps.append(op.utc_time)
+
+        while len(timestamps) > 0:
+            timestamp = timestamps.pop(0)
+
+            if len(timestamppairs) > 0 and timestamppairs[-1][0]+datetime.timedelta(minutes=995) > timestamp:
```

Comment:
1. yes true the function as a whole tries to group the timestamps into ranges below 995 minutes 995 because the current binance max is 1000minutes and i want to be able to get 5 candles after the last timestamp( this could be reduced to 1 candle) i haven't looked into what the max is for other exchanges is and how fetch_ohlcv handles stuff with a larger range than the exchange api allows.
2. good point
3. yeah that looks better than my implementation